src/lora_subnet.erl

```diff
@@ -0,0 +1,536 @@
+    DevAddr.
+
+-spec subnet_addr_to_netid(subnetaddr(), [netid()]) -> netid().
+subnet_addr_to_netid(NwkAddr, NetIDList) ->
+    subnet_addr_to_netid_search(NwkAddr, NetIDList, NetIDList).
+
+subnet_addr_to_netid_search(_, [], _) ->
+    0;
+subnet_addr_to_netid_search(NwkAddr, SList, NetIDList) ->
+    [H | T] = SList,
+    WithinRange = subnet_addr_within_range(NwkAddr, H, NetIDList),
+    case WithinRange of
+        true ->
+            H;
+        false ->
+            subnet_addr_to_netid_search(NwkAddr, T, NetIDList)
+    end.
+
+-spec subnet_addr_within_range(subnetaddr(), netid(), [netid()]) -> boolean().
+subnet_addr_within_range(Addr, NetID, NetIDList) ->
+    {Lower, Upper} = netid_addr_range(NetID, NetIDList),
+    (Addr >= Lower) and (Addr < Upper).
+
+-spec var_net_class(netclass()) -> non_neg_integer().
+var_net_class(NetClass) when NetClass =< 7 ->
+    IDLen = id_len(NetClass),
+    case NetClass of
+        0 -> 0;
+        1 -> 2#10 bsl IDLen;
+        2 -> 2#110 bsl IDLen;
+        3 -> 2#1110 bsl IDLen;
+        4 -> 2#11110 bsl IDLen;
+        5 -> 2#111110 bsl IDLen;
+        6 -> 2#1111110 bsl IDLen;
+        7 -> 2#11111110 bsl IDLen
+    end.
+
+-spec var_netid(netclass(), netid()) -> non_neg_integer().
+var_netid(NetClass, NetID) when NetClass =< 7 ->
+    NetID bsl addr_len(NetClass).
+
+-spec the_netid(number() | binary()) -> netid().
+the_netid(DevNum) when erlang:is_number(DevNum) ->
+    the_netid(<<DevNum:32/integer-unsigned>>);
+the_netid(DevAddr) ->
+    Type = netid_type(DevAddr),
+    ID = parse_id(DevAddr, Type + 1, id_len(Type)),
+    NetID = ID bor (Type bsl 21),
+    NetID.
+
+-spec parse_netid(number() | binary()) -> {ok, netid()} | {error, invalid_netid_type}.
+parse_netid(DevNum) when erlang:is_number(DevNum) ->
+    parse_netid(<<DevNum:32/integer-unsigned>>);
+parse_netid(DevAddr) ->
+    try
+        Type = netid_type(DevAddr),
+        ID = parse_id(DevAddr, Type + 1, id_len(Type)),
+        {ok, ID bor (Type bsl 21)}
+    catch
+        throw:invalid_netid_type:_ ->
+            {error, invalid_netid_type}
+    end.
+
+-spec addr_bit_len(number() | binary()) -> 7 | 10 | 13 | 15 | 17 | 20 | 24 | 25.
+addr_bit_len(DevNum) when erlang:is_number(DevNum) ->
+    addr_bit_len(<<DevNum:32/integer-unsigned>>);
+addr_bit_len(DevAddr) ->
+    NetID = the_netid(DevAddr),
+    addr_len(netid_class(NetID)).
+
+-spec parse_id(binary(), non_neg_integer(), non_neg_integer()) -> netid().
+parse_id(DevAddr, PrefixLength, NwkIDBits) ->
+    <<DevAddrNum:32/integer-unsigned>> = DevAddr,
+    uint32(uint32(DevAddrNum bsl PrefixLength) bsr (32 - NwkIDBits)).
+
+-spec netid_type(number() | binary()) -> 0..7.
+netid_type(NetID) when erlang:is_number(NetID) ->
+    netid_type(<<NetID:32/integer-unsigned>>);
+netid_type(<<First:8/integer-unsigned, _/binary>>) ->
+    netid_type(First, 7).
+
+-spec netid_type(non_neg_integer(), non_neg_integer()) -> 0..7.
+netid_type(_, -1) ->
+    throw(invalid_netid_type);
```

Comment:
Please highlight this side-effect in the README: `throw()` will be unexpected in a functional programming context.

I'm all for "Let it crash", but it's important to know in what ways it will do so.  To that end, I often annotate such caveats in docs or API guides with the statement:
*"Let it crash" means revert to a well-defined, stable state even if that means killing the worker process and having the Supervisor restart it.*